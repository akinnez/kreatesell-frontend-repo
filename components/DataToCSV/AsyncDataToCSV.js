import React from "react";
import { Button } from "antd";
import axios from "axios";
import { CSVLink } from "react-csv";
import { BsDownload } from "react-icons/bs";
import { showToast } from "utils";
import styles from "./index.module.scss";

/*  AsyncDataToCSV Component receives 3.props:
  1 - this.props.url = link to the resource, an array of data, to be converted to downloadable CSV

  2 - this.props.headers = The CSV file headers. Format below:
  const headers = [
    { label: "Product Name", key: "product_name" },
    { label: "Product Type", key: "product_type" },
    { label: "Product Price", key: "product_price" },
    { label: "Date Created", key: "date_created" },
  ];

  3 - this.props.filename = The default name of the csv file.
*/

export default class AsyncDataToCSV extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      data: [],
      loading: false,
    };

    this.csvRef = React.createRef();
  }

  getData = async link => {
    try {
      this.setState({ loading: true });

      const response = await axios.get(link);
      return response.data;
    } catch {
      return null;
    } finally {
      this.setState({ loading: false });
    }
  };

  downloadCSV = async () => {
    const data = await this.getData(this.props.url);

    if (!data) {
      showToast("Could not download CSV. Try again later", "error");
      return;
    }

    this.setState({ data: data.data }, () => {
      setTimeout(() => {
        this.csvRef.current.link.click();
      });
    });
  };

  render() {
    const { data, loading } = this.state;
    const { headers, filename } = this.props;

    return (
      <>
        <div className={styles.csvDataDownload}>
          <Button type="link" onClick={this.downloadCSV} loading={loading}>
            Export data in CSV &nbsp;
            <BsDownload />
          </Button>
        </div>
        <CSVLink
          headers={headers}
          filename={`${filename}.csv`}
          data={data}
          ref={this.csvRef}
        />
      </>
    );
  }
}
