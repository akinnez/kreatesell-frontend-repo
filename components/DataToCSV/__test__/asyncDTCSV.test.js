import renderer from 'react-test-renderer';
import AsyncDataToCSV from '../AsyncDataToCSV';

it('renders a snapshot of <AsyncDataToCSV />', () => {
	const toCSV = renderer.create(<AsyncDataToCSV />).toJSON();
	expect(toCSV).toMatchSnapshot();
});
