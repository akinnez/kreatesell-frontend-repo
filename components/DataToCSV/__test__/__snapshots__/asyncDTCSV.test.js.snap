// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot of <AsyncDataToCSV /> 1`] = `
Array [
  <div
    className="csvDataDownload"
  >
    <button
      className="ant-btn ant-btn-link"
      onClick={[Function]}
      type="button"
    >
      <span>
        Export data in CSV  
      </span>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M.5 9.9a.5.5 0 0 1 .5.5v2.5a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1v-2.5a.5.5 0 0 1 1 0v2.5a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2v-2.5a.5.5 0 0 1 .5-.5z"
        />
        <path
          d="M7.646 11.854a.5.5 0 0 0 .708 0l3-3a.5.5 0 0 0-.708-.708L8.5 10.293V1.5a.5.5 0 0 0-1 0v8.793L5.354 8.146a.5.5 0 1 0-.708.708l3 3z"
        />
      </svg>
    </button>
  </div>,
  <a
    download="undefined.csv"
    href="data:text/csv;charset=utf-8,﻿"
    onClick={[Function]}
    target="_self"
  />,
]
`;
